libraries/chain/include/eosio/chain/merkle.hpp

```diff
@@ -1,29 +1,82 @@
 #pragma once
 #include <eosio/chain/types.hpp>
+#include <fc/io/raw.hpp>
+#include <bit>
+#include <array>
+#include <future>
 
-namespace eosio { namespace chain {
+namespace eosio::chain {
 
-   digest_type make_canonical_left(const digest_type& val);
-   digest_type make_canonical_right(const digest_type& val);
+namespace detail {
 
-   bool is_canonical_left(const digest_type& val);
-   bool is_canonical_right(const digest_type& val);
+#if __cplusplus >= 202002L
+   inline int      popcount(uint64_t x)  noexcept { return std::popcount(x); }
+   inline uint64_t bit_floor(uint64_t x) noexcept { return std::bit_floor(x); }
+#else
+   inline int      popcount(uint64_t x)  noexcept { return __builtin_popcountll(x); }
+   inline uint64_t bit_floor(uint64_t x) noexcept { return x == 0 ? 0ull : 1ull << (64 - 1 - __builtin_clzll(x)); }
+#endif
 
+inline digest_type hash_combine(const digest_type& a, const digest_type& b) {
+   return digest_type::hash(std::make_pair(std::cref(a), std::cref(b)));
+}
 
-   inline auto make_canonical_pair(const digest_type& l, const digest_type& r) {
-      return make_pair(make_canonical_left(l), make_canonical_right(r));
-   };
+// does not overwrite passed sequence
+//
+// log2 recursion OK, uses less than 5KB stack space for 4 billion digests
+// appended (or 0.25% of default 2MB thread stack size on Ubuntu).
+// -----------------------------------------------------------------------
+template <class It, bool async = false>
+requires std::is_same_v<std::decay_t<typename std::iterator_traits<It>::value_type>, digest_type>
+inline digest_type calculate_merkle_pow2(const It& start, const It& end) {
+   auto size = end - start;
+   assert(size >= 2);
+   assert(detail::bit_floor(static_cast<size_t>(size)) == size);
 
-   /**
-    *  Calculates the merkle root of a set of digests, if ids is odd it will duplicate the last id.
-    *  Uses make_canonical_pair which before hashing sets the first bit of the previous hashes
-    *  to 0 or 1 to indicate the side it is on.
-    */
-   digest_type legacy_merkle( deque<digest_type>&& ids );
+   if (size == 2)
+      return hash_combine(start[0], start[1]);
+   else {
+      if (async && size >= 4096) {                    // below 4096, starting async threads is overkill
+         std::array<std::future<digest_type>, 4> fut; // size dictates the number of threads (must be power of two)
```

Comment:
This is likely fine as it is only 4 threads. We do need to be mindful of thread usage as we don't want to overwhelm a system so that it slows down the main thread. Other thread usage is setup at startup. Would be nice to be able to provide an `io_context` for these tasks so they could run on the chain thread pool; however likely not worth the effort.

Reply:
I updated it so that it uses only 2 threads up to 4096 digests, and 4 for 4096 and up.